include/bluetooth/gatt.h

```diff
@@ -1669,6 +1669,13 @@ struct bt_gatt_subscribe_params {
 #endif /* CONFIG_BT_GATT_AUTO_DISCOVER_CCC */
 	/** Subscribe value */
 	uint16_t value;
+#if defined(CONFIG_BT_SMP)
+	/** Minimum required security for received notification. If not set
+	 * explicitly by application it is set by host based on security level
+	 * at the time of subscription.
+	 */
+	bt_security_t min_security;
```

Comment:
Did you verify whether there are any cases in Zephyr that doesn't use a zero-allocated version of this struct? We can't control other applications, but at least we can ensure that in the Zephyr repo that no use of `bt_gatt_subscribe_params` has this value uninitialized.

Reply:
We could/should also update existing uses in Zephyr to set this to the security level of the CCCD that we are subscribing to. Having that in e.g. the sample would also show how this field should be used.

Reply:
I've assumed that since subscribe params needs to be valid for time of subscription those are static (and thus initialized) but if one use heap that might not be the case..

maybe it should be opt-in via flag as in initial version of this PR?

Reply:
I didn't see the initial version of this PR, so I don't know :)

Reply:
I've pushed original version to another branch for reference
https://github.com/sjanc/zephyr/tree/gatt_client_sec2

Reply:
Thanks. 

The issue with the flag, is that it conflicts with the "shall" requirements in the core spec, so it cannot be optional. This also leaves me to believe that having this value, as an application controlled value, isn't sufficient either. Since the client _shall_ check the configuration, having an application setting this value isn't strong enough. Based on that, I think we need to use the security requirements of the CCCD to set such a value (as suggested in https://github.com/zephyrproject-rtos/zephyr/pull/42726#discussion_r806797099 (my comment below)). 

Do you agree/disagree with that? 

@asbjornsabo You are very well versed in reading and understanding the spec; what is your take on this?

Reply:
I'm fine with either opt-in or opt-out but this is always application that decides if security is required, whenever explicitly by setting flag (and/or level) or implicitly by subscribing only after required sec level is reached

Reply:
so how do you want to figure out which security level is required minimum if not from application implementing profile?

I agree that it is GATT server responsibility to not send notification if security level is too low but this PR is about protecting from rogue device

Reply:
> I agree that it is GATT server responsibility to not send notification if security level is too low but this PR is about protecting from rogue device

I agree, but the comment was regarding what the minimum security level should be, and that it shouldn't be below the security level of what the server has set, otherwise it doesn't provide anything. E.g. if the CCCD has a security level of L3, then the application should not set the `min_security` to L1, as that would not protect against any rogue devices. Furthermore, if the setting of `min_security` is the responsible of the application, then the application might as well do this check and discard, instead of the stack. 

Though, given that the spec states a "shall" for the GATT client, this should, IMO, reside in the stack. 
Do you have a name/reference to the test case that tests this? Does it mention an upper tester? 

> so how do you want to figure out which security level is required minimum if not from application implementing profile?

That's a good question. It isn't possible for the GATT client to know the permission beforehand, except the minimum value expressed in the service specification. It isn't clear to me how this requirement was ever supposed to be handled by the client without exposing that information to the client, except using the service specification. 

Possibly this solution is the only one that exists, unless we always just default to the current connection, but that is also a poor design. 

For something like the "Service Changed" characteristic, we certainly aren't interested in ever discarding that notification.

Reply:
Affected tests are mentioned in PR description

It is up to profile not GATT client to decide about required security,  not sure how you want to handle that in stack without info provided by application. Profile/service define security of own services, not GATT.

And yes, it is up to application to decide if notification should or should not be processed,  this PR is just to make this a common check and give applications option to not have to handle this in own code for every subscription.

Reply:
You've convinced me that this is the only way to do this properly :)

Thanks for your input @sjanc